### DISCUSSION TITLE
performance problems with cached queries

### DISCUSSION THREAD

**Discussion by eltoder** [sqlalchemy/sqlalchemy#11108] [Usage Questions]
### Ensure stubs packages are not installed

- [X] No sqlalchemy stub packages is installed (both `sqlalchemy-stubs` and `sqlalchemy2-stubs` are not compatible with v2)

### Verify if the api is typed

- [X] The api is not in a module listed in [#6810](https://github.com/sqlalchemy/sqlalchemy/issues/6810) so it should pass type checking

### Describe the typing issue

Passing `stmt.compile()` to `execute()` generates type errors in mypy, but appears to work fine. Should the result of `compile` be marked as `Executable`?

### To Reproduce

```python
import sqlalchemy as sa

engine = sa.create_engine("sqlite:///:memory:")
stmt = sa.select(sa.literal_column("1", type_=sa.Integer))
compiled: sa.sql.compiler.Compiled = stmt.compile(engine)
with engine.connect() as conn:
    print(conn.execute(compiled).all())
    print(conn.execute(stmt.compile(engine)).all())
```

### Error

```
compile_exec.py:7: error: No overload variant of "execute" of "Connection" matches argument type "Compiled"  [call-overload]
compile_exec.py:7: note: Possible overload variants:
compile_exec.py:7: note:     def [_T] execute(self, statement: TypedReturnsRows[_T], parameters: Sequence[Mapping[str, Any]] | Mapping[str, Any] | None = ..., *, execution_options: _CoreKnownExecutionOptions | Mapping[str, Any] | None = ...) -> CursorResult[_T]
compile_exec.py:7: note:     def execute(self, statement: Executable, parameters: Sequence[Mapping[str, Any]] | Mapping[str, Any] | None = ..., *, execution_options: _CoreKnownExecutionOptions | Mapping[str, Any] | None = ...) -> CursorResult[Any]
compile_exec.py:8: error: No overload variant of "execute" of "Connection" matches argument type "SQLCompiler"  [call-overload]
compile_exec.py:8: note: Possible overload variants:
compile_exec.py:8: note:     def [_T] execute(self, statement: TypedReturnsRows[_T], parameters: Sequence[Mapping[str, Any]] | Mapping[str, Any] | None = ..., *, execution_options: _CoreKnownExecutionOptions | Mapping[str, Any] | None = ...) -> CursorResult[_T]
compile_exec.py:8: note:     def execute(self, statement: Executable, parameters: Sequence[Mapping[str, Any]] | Mapping[str, Any] | None = ..., *, execution_options: _CoreKnownExecutionOptions | Mapping[str, Any] | None = ...) -> CursorResult[Any]
Found 2 errors in 1 file (checked 1 source file)
```

### Versions

- OS: Linux
- Python: 3.10.13
- SQLAlchemy: 2.0.28
- Type checker: mypy 1.8.0 (compiled: yes)

**Comment by zzzeek** [comment]
hi -

why are you doing this?

```
print(conn.execute(stmt.compile(engine)).all())
```

that's pretty legacy stuff and I'd rather just remove it in 2.1.  is there some use case you have?

**Comment by eltoder** [comment]
I'm caching Compiled objects. I know that engine already has a cache, but I observed a 10-20% speedup from caching Compiled objects versus caching statements.

**Comment by zzzeek** [comment]
there's going to be problems using compile() like that since there's lots of arguments that you arent sending there which would be needed for some kinds of statements

also try taking your select() statement and using the same actual Python object each time.  the cache key generation overhead which is what you are seeing will go to zero.    Or use a https://docs.sqlalchemy.org/en/20/core/connections.html#engine-lambda-caching .  that will also remove the caching overhead

**Reply by eltoder** [reply]
Could you elaborate what am I not passing? This is covered by unit tests: https://github.com/sqlalchemy/sqlalchemy/blob/1295731364965e81a4ab76415e7baa3137b68917/test/sql/test_query.py#L274
It works for everything I tried so far.

I tried lambda_stmt, but found the style awkward. It is easier to cache select statements. I am using the same actual object, since this is what was cached, but still caching compiled objects is faster. I'll try to write a standalone benchmark.

**Reply by zzzeek** [reply]
take a look

https://github.com/sqlalchemy/sqlalchemy/blob/1295731364965e81a4ab76415e7baa3137b68917/lib/sqlalchemy/engine/base.py#L1639-L1658

we dont want to support a large public API for all those options, we need to be able to add new keys at any time to that internal.  it's a burden here for stmt.compile() with no args to support all that in an execution context.   by us supporting execute(compiled), now someone might come along and say "my schema_translate_map is not working when I run execute(compiled)" ; we say, "well that requires more things to be passed along" and they say "it should just work / be documented / etc", then repeat that for every other key in there.

you're of course free to just vendor `_execute_compiled()` and do your own thing.

**Reply by zzzeek** [reply]
also as I said, if you use the same select() object each time, the cache key overhead goes to zero.   see test_core_reuse_stmt at https://docs.sqlalchemy.org/en/20/_modules/examples/performance/short_selects.html
